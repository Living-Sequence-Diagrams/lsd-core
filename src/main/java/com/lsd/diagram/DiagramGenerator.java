package com.lsd.diagram;

import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Participant;
import com.lsd.report.model.SequenceDiagram;
import com.lsd.events.SequenceEvent;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<SequenceEvent> events;

    public SequenceDiagram sequenceDiagram() {
        String uml = generateSequenceUml();
        String svg = generateSequenceSvg(uml);
        return SequenceDiagram.builder()
                .uml(uml)
                .svg(svg)
                .build();
    }

    private String generateSequenceSvg(String markup) {
        SvgConverter svgConverter = new SvgConverter();

        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(SequenceEvent::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
