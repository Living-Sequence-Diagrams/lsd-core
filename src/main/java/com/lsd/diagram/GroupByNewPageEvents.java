package com.lsd.diagram;

import com.lsd.events.Newpage;
import com.lsd.events.PageTitle;
import com.lsd.events.SequenceEvent;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.function.BiConsumer;
import java.util.function.BinaryOperator;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Collector;

/**
 * Splits a given List of sequence events into sublists if NewPage events are found. The NewPage event is converted to
 * a PageTitle event (preserving the title).
 */
class GroupByNewPageEvents implements Collector<SequenceEvent, List<List<SequenceEvent>>, List<List<SequenceEvent>>> {
    @Override
    public Supplier<List<List<SequenceEvent>>> supplier() {
        return () -> {
            var lists = new ArrayList<List<SequenceEvent>>();
            lists.add(new ArrayList<>());
            return lists;
        };
    }

    @Override
    public BiConsumer<List<List<SequenceEvent>>, SequenceEvent> accumulator() {
        return (lists, sequenceEvent) -> {
            if (sequenceEvent instanceof Newpage) {
                var newGrouping = new ArrayList<SequenceEvent>();
                newGrouping.add(PageTitle.titled(((Newpage) sequenceEvent).getTitle()));
                lists.add(newGrouping);
            } else {
                var currentGrouping = lists.get(lists.size() - 1);
                currentGrouping.add(sequenceEvent);
            }
        };
    }

    @Override
    public BinaryOperator<List<List<SequenceEvent>>> combiner() {
        return null;
    }

    @Override
    public Function<List<List<SequenceEvent>>, List<List<SequenceEvent>>> finisher() {
        return lists -> lists;
    }

    @Override
    public Set<Characteristics> characteristics() {
        return Set.of();
    }
}
