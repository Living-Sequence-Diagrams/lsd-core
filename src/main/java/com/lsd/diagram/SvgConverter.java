package com.lsd.diagram;

import net.sourceforge.plantuml.FileFormatOption;
import net.sourceforge.plantuml.SourceStringReader;
import org.jdom.Document;
import org.jdom.JDOMException;
import org.jdom.input.SAXBuilder;
import org.jdom.output.Format;
import org.jdom.output.XMLOutputter;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;

import static net.sourceforge.plantuml.FileFormat.SVG;

public class SvgConverter {

    public String convert(String markup) {
        return prettyPrint(createSvg(markup));
    }

    private String createSvg(String plantUmlMarkup) {
        SourceStringReader reader = new SourceStringReader(plantUmlMarkup);
        try (final ByteArrayOutputStream os = new ByteArrayOutputStream()) {
            reader.outputImage(os, new FileFormatOption(SVG, false));
            return os.toString();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private String prettyPrint(String xml) {
        try {
            SAXBuilder sb = new SAXBuilder();
            Document doc = sb.build(new StringReader(xml));
            StringWriter stringWriter = new StringWriter();
            new XMLOutputter(Format.getPrettyFormat()).output(doc, stringWriter);
            return stringWriter.toString();
        } catch (JDOMException | IOException e) {
            throw new RuntimeException(e);
        }
    }
}