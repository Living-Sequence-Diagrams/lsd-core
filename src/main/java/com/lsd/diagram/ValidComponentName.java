package com.lsd.diagram;

import org.apache.commons.text.WordUtils;

import static java.util.Arrays.stream;
import static java.util.stream.Collectors.joining;

public class ValidComponentName {
    public static final String ILLEGAL_CHARS = "[()/]";

    public static String of(String name) {
        return stream(name
                .replaceAll(ILLEGAL_CHARS, " ")
                .split(" ")
        ).map(WordUtils::capitalize)
                .collect(joining())
                .replaceAll(" ", "");
    }
}
