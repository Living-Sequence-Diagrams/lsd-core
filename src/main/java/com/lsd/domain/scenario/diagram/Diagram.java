package com.lsd.domain.scenario.diagram;

public interface Diagram {

    /**
     * @return Returns an SVG representation of the diagram
     */
    String toSvg();
}
