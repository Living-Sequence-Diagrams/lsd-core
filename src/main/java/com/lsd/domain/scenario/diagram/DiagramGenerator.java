package com.lsd.domain.scenario.diagram;

import com.lsd.domain.Participant;
import com.lsd.domain.scenario.events.Event;
import com.lsd.domain.scenario.events.Interaction;
import com.lsd.properties.LsdProperties;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<Event> events;

    public SequenceDiagram sequenceDiagram() {
        return SequenceDiagram.builder()
                .uml(generateSequenceUml())
                .interactions(events.stream()
                        .filter(Interaction.class::isInstance)
                        .map(Interaction.class::cast)
                        .collect(toList()))
                .build();
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(Event::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
