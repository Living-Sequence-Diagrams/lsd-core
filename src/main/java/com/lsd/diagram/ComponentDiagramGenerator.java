package com.lsd.diagram;

import com.lsd.IdGenerator;
import com.lsd.events.Message;
import com.lsd.events.SequenceEvent;
import com.lsd.events.SynchronousResponse;
import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Diagram;
import com.lsd.report.model.Participant;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.function.Predicate.not;
import static java.util.stream.Collectors.toList;

@Builder
public class ComponentDiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/component-uml.peb");
    private final SvgConverter svgConverter = new SvgConverter();

    private final List<Participant> participants;
    private final List<SequenceEvent> events;
    private final IdGenerator idGenerator;

    public Diagram diagram() {
        String uml = generateUml();
        return Diagram.builder()
                .id(idGenerator.next())
                .uml(uml)
                .svg(generateSvg(uml))
                .build();
    }

    private String generateSvg(String markup) {
        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "participants", participants.stream()
                        .map(Participant::getMarkup)
                        .map(markup -> markup.replaceAll("^participant ", "component "))
                        .distinct()
                        .collect(toList()),
                "events", events.stream()
                        .filter(event -> event instanceof Message)
                        .filter(not(SynchronousResponse.class::isInstance))
                        .map(event -> (Message) event)
                        .map(message -> String.format("[%s] -> %s", message.getFrom(), message.getTo()))
                        .distinct()
                        .collect(toList())
        ));
        return writer.toString();
    }
}
