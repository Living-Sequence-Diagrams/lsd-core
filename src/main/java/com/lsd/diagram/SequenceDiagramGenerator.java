package com.lsd.diagram;

import com.lsd.IdGenerator;
import com.lsd.events.SequenceEvent;
import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Diagram;
import com.lsd.report.model.Participant;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;
import static org.apache.commons.collections4.ListUtils.partition;

@Builder
public class SequenceDiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<SequenceEvent> events;
    private final IdGenerator idGenerator;

    public Optional<Diagram> diagram(int maxEventsPerDiagram) {
        if (events.isEmpty()) {
            return Optional.empty();
        }
        var uml = partition(events, maxEventsPerDiagram).stream()
                .map(this::generateSequenceUml)
                .collect(joining(System.lineSeparator()));

        String svg = generateSequenceSvg(uml);
        return Optional.of(Diagram.builder()
                .id(idGenerator.next())
                .uml(uml)
                .svg(svg)
                .build());
    }

    private String generateSequenceSvg(String markup) {
        SvgConverter svgConverter = new SvgConverter();

        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateSequenceUml(List<SequenceEvent> events) {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants.stream().distinct().collect(toList()),
                "events", events.stream()
                        .map(SequenceEvent::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
