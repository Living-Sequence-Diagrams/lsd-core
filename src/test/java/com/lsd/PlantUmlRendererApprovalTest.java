package com.lsd;

import com.lsd.domain.scenario.diagram.DiagramGenerator;
import com.lsd.domain.scenario.events.SimpleInteraction;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Set;

class PlantUmlRendererApprovalTest {

    private DiagramGenerator diagramGenerator = DiagramGenerator.builder()
            .participants(List.of())
            .includes(Set.of())
            .events(List.of(
                    SimpleInteraction.builder()
                            .id("1")
                            .label("Sending a request")
                            .from("A")
                            .to("B")
                            .data("Please help?")
                            .build(),
                    SimpleInteraction.builder()
                            .id("2")
                            .label("Sending a response")
                            .from("B")
                            .to("A")
                            .data("OK")
                            .build()
            )).build();

    @Test
    void renderSequenceUml() {
        Approvals.verify(diagramGenerator.sequenceDiagram().getUml());
    }

    @Test
    void renderSequenceSvg() {
        Approvals.verifyXml(diagramGenerator.sequenceDiagram().toSvg());
    }
}