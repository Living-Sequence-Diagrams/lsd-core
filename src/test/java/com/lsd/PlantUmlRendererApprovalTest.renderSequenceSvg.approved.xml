<?xml version="1.0" encoding="UTF-8"?><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentScriptType="application/ecmascript" contentStyleType="text/css" height="162px" preserveAspectRatio="none" style="width:188px;height:162px;" version="1.1" viewBox="0 0 188 162" width="188px" zoomAndPan="magnify">
    
  <defs>
        
    <filter height="300%" id="favoohugmnn2p" width="300%" x="-1" y="-1">
            
      <feGaussianBlur result="blurOut" stdDeviation="2.0"/>
            
      <feColorMatrix in="blurOut" result="blurOut2" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 .4 0"/>
            
      <feOffset dx="4.0" dy="4.0" in="blurOut2" result="blurOut3"/>
            
      <feBlend in="SourceGraphic" in2="blurOut3" mode="normal"/>
          
    </filter>
      
  </defs>
    
  <g>
        
    <line style="stroke:#A80036;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="19" x2="19" y1="40.4883" y2="119.1094"/>
        
    <line style="stroke:#A80036;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="165" x2="165" y1="40.4883" y2="119.1094"/>
        
    <rect fill="#FEFECE" filter="url(#favoohugmnn2p)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="24" x="5" y="5"/>
        
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="10" x="12" y="25.5352">A</text>
        
    <rect fill="#FEFECE" filter="url(#favoohugmnn2p)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="24" x="5" y="118.1094"/>
        
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="10" x="12" y="138.6445">A</text>
        
    <rect fill="#FEFECE" filter="url(#favoohugmnn2p)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="22" x="152" y="5"/>
        
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="8" x="159" y="25.5352">B</text>
        
    <rect fill="#FEFECE" filter="url(#favoohugmnn2p)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="22" x="152" y="118.1094"/>
        
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="8" x="159" y="138.6445">B</text>
        
    <polygon fill="#A80036" points="153,67.7988,163,71.7988,153,75.7988,157,71.7988" style="stroke:#A80036;stroke-width:1.0;"/>
        
    <line style="stroke:#A80036;stroke-width:1.0;" x1="19" x2="159" y1="71.7988" y2="71.7988"/>
        
    <a href="#1" target="_top" title="#1" xlink:actuate="onRequest" xlink:href="#1" xlink:show="new" xlink:title="#1" xlink:type="simple">
            
      <text fill="#0000FF" font-family="sans-serif" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="112" x="26" y="67.0566">Sending a request</text>
          
    </a>
        
    <polygon fill="#A80036" points="30,97.1094,20,101.1094,30,105.1094,26,101.1094" style="stroke:#A80036;stroke-width:1.0;"/>
        
    <line style="stroke:#A80036;stroke-width:1.0;" x1="24" x2="164" y1="101.1094" y2="101.1094"/>
        
    <a href="#2" target="_top" title="#2" xlink:actuate="onRequest" xlink:href="#2" xlink:show="new" xlink:title="#2" xlink:type="simple">
            
      <text fill="#0000FF" font-family="sans-serif" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="122" x="36" y="96.3672">Sending a response</text>
          
    </a>
      
  </g>
  
</svg>
